x-pack/plugins/entity_manager/server/lib/v2/definitions/source_definition.ts

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { IScopedClusterClient, Logger } from '@kbn/core/server';
+import { DEFINITIONS_ALIAS, TEMPLATE_VERSION } from '../constants';
+import {
+  EntitySourceDefinition,
+  CreateOperationStatus,
+  ReadOperationStatus,
+  StoredEntitySourceDefinition,
+} from '../types';
+import { runESQLQuery } from '../run_esql_query';
+
+export async function storeSourceDefinition(
+  source: EntitySourceDefinition,
+  clusterClient: IScopedClusterClient,
+  logger: Logger
+): Promise<CreateOperationStatus<EntitySourceDefinition>> {
+  try {
+    const esClient = clusterClient.asInternalUser;
+
+    const sources = await runESQLQuery('fetch source definition for conflict check', {
+      esClient,
+      query: `FROM ${DEFINITIONS_ALIAS} METADATA _id | WHERE definition_type == "source" AND _id == "source:${source.id}"`,
+      logger,
+    });
+
+    if (sources.length !== 0) {
+      return {
+        status: 'conflict',
+        reason: `An entity source definition with the ID "${source.id}" already exists.`,
+      };
+    }
+
+    const definition: StoredEntitySourceDefinition = {
+      template_version: TEMPLATE_VERSION,
+      definition_type: 'source',
+      source,
+    };
+
+    await esClient.index({
+      index: DEFINITIONS_ALIAS,
+      id: `source:${definition.source.id}`,
+      document: definition,
+    });
+
+    return {
+      status: 'success',
+      resource: definition.source,
+    };
+  } catch (error) {
+    return {
+      status: 'error',
+      reason: `${error.name}: ${error.message}`,
+    };
```

Comment:
why not throw ? this makes the entityClient interface inconsistent across its methods as some will throw and others don't. we also lose the stacktrace along the way

Reply:
I can't recall, I think I added this while working on bulk support to make it easier to reduce the result.
I can't see a good reason to keep it now so I'll replace it with throwing and use a custom error for the conflict case.